### PR TITLE
Add support for deflate and snappy codecs

### PR DESCRIPTION
**Pull request by smalot**
Currently, only `null` codec is supported.
With Google tools, the `snappy` codec is often used.
http://google.github.io/snappy/

For `deflate` codec, I applied this patch : https://github.com/apache/avro/pull/561/commits/8a218955d0c1a197822db80b2065de1725cb04dd
For `snappy` codec, I used the `snappy` php extension in addition with a crc32 checksum.

### CODE REVIEW

**Comment by smalot** [comment]
Hi @tPl0ch 
Could you double check this PR ?

File: test/data/users.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"name":"Alyssa","favorite_color":null,"favorite_numbers":[3,9,15,20]}
```
Comment (tPl0ch) [inline test/data/users.json:1]: Should have a newline at end of file

File: test/DataFileTest.php [hunk 2]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
-    $data_file = $this->add_data_file('data-wr-nothing-null.avr');
-    $writers_schema = '"null"';
-    $dw = AvroDataIO::open_file($data_file, 'w', $writers_schema);
-    $dw->close();
-
-    $dr = AvroDataIO::open_file($data_file);
-    $data = $dr->data();
-    $dr->close();
-    $this->assertEmpty($data);
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:79]: Should be refactored into a PHPUnit data provider and used via the `@dataProvider` annotation.

File: test/DataFileTest.php [hunk 3]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
-    $dw = AvroDataIO::open_file($data_file, 'w', $writers_schema);
-    $dw->append($data);
-    $dw->close();
-
-    $dr = AvroDataIO::open_file($data_file);
-    $read_data = $dr->data();
-    $read_data = reset($read_data);
-    $dr->close();
-    $this->assertEquals($data, $read_data);
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:94]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 4]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
```
Comment (tPl0ch) [inline test/DataFileTest.php:112]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 5]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
-    $dw->append(1);
-    $dw->close();
-
-    $dr = AvroDataIO::open_file($data_file);
-    $read_data = $dr->data();
-    $read_data = reset($read_data);
-    $dr->close();
-    $this->assertEquals($data, $read_data);
-
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:130]: Can re-use above data provider for codecs.

File: test/DataFileTest.php (hunk 4 above)
Comment (tPl0ch) [inline test/DataFileTest.php:112]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 6]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
 
   public function test_write_read_true_round_trip()
   {
-    $data_file = $this->add_data_file('data-wr-true.avr');
-    $writers_schema = '"boolean"';
-    $datum = true;
-    $dw = AvroDataIO::open_file($data_file, 'w', $writers_schema);
-    $dw->append($datum);
-    $dw->close();
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:149]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 7]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
 
   public function test_write_read_false_round_trip()
   {
-    $data_file = $this->add_data_file('data-wr-false.avr');
-    $writers_schema = '"boolean"';
-    $datum = false;
-    $dw = AvroDataIO::open_file($data_file, 'w', $writers_schema);
-    $dw->append($datum);
-    $dw->close();
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:167]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 8]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
-      $dw->append($datum);
-    $dw->close();
-
-    $dr = AvroDataIO::open_file($data_file);
-    $read_data = $dr->data();
-    $dr->close();
-    $this->assertEquals($data, $read_data,
-                        sprintf("in: %s\nout: %s",
-                                json_encode($data), json_encode($read_data)));
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:185]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 9]
```diff
@@ -68,130 +68,144 @@ protected function setUp()
...
+      $this->assertEquals($data, $read_data,
+                          sprintf("in: %s\nout: %s",
+                                  json_encode($data), json_encode($read_data)));
+    }
   }
 
   public function test_differing_schemas_with_primitives()
   {
-    $data_file = $this->add_data_file('data-prim.avr');
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:205]: Can re-use above data provider for codecs.

File: test/DataFileTest.php [hunk 10]
```diff
@@ -200,33 +214,33 @@ public function test_differing_schemas_with_primitives()
...
+      foreach ($dr->data() as $index => $record) {
+        $this->assertEquals($data[$index]['username'], $record['username']);
+      }
     }
   }
 
   public function test_differing_schemas_with_complex_objects()
   {
-    $data_file = $this->add_data_file('data-complex.avr');
+    foreach (AvroDataIO::valid_codecs() as $codec) {
```
Comment (tPl0ch) [inline test/DataFileTest.php:240]: Can re-use above data provider for codecs.

**Review by tPl0ch — CHANGES_REQUESTED** [review]
Thanks, there are just a few minor comments - mainly getting rid of the iterations in the tests via a data provider.

File: test/data/users.json (hunk 1 above)
Comment (smalot) [reply test/data/users.json:1]: Indeed

File: test/DataFileTest.php (hunk 2 above)
Comment (smalot) [reply test/DataFileTest.php:79]: Not sure to understand what you want.
Can you provide me an example ?

Comment (smalot) [reply test/DataFileTest.php:79]: I think I understand
I'll try something

https://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

File: test/data/users.json (hunk 1 above)
Comment (smalot) [reply test/data/users.json]: Fixed

**Comment by smalot** [comment]
@tPl0ch  Should be good, can you check on your side ?

**Review by tPl0ch — APPROVED** [review]
LGTM

**Comment by smalot** [comment]
Thank you @tPl0ch

**Pull request merged**